### ISSUE TITLE
Ehhancement: numbers in json

### ISSUE THREAD

**Issue by faxm0dem** [balabit/syslog-ng#2219]
It would be awesome if numbers were supported in json-format.
Our use-case is for Elasticsearch. If syslog-ng supported numbers in the json output, it would greatly simplify the Elasticsearch side, where we need to creatre special mappings for the numbers which appear as strings coming from syslog-ng.

**Comment by algernon** [comment]
In an arguably limited way, it already does: you can give type-hints to value-pairs, and `format-json` will honor those. The admin guide has a [short section](https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.16/administration-guide/9#TOPIC-956418) on this. The short story is that you *can* specify a type on a per-field basis. There's no automatism to do this, and you have to list every single property you wish to type-hint, without the being able to use wildcards. But if you have a limited set of fields, this can work.

**Comment by faxm0dem** [comment]
Thanks, I didn't know this was already implemented.
That being said, it would be an option if patterndb and json-parser supported type-hints *and* passed them automatically to format-json. Is that the case? If not, I shall rename this issue to ask for that scenario.

Cheers

**Comment by algernon** [comment]
No, neither patterndb nor json-parser support type hints at the moment. It was one of the suggested projects during [GSoC2014](https://github.com/balabit/syslog-ng/wiki/GSoC2014-idea-&-project-list#type-hinting-from-the-start), but the project did not get selected. As far as I know, no effort was made since then towards this goal.

**Comment by bazsi** [comment]
Patterndb could have a way to record types, we have a byte reserved for
this purpose in the LogMessage structure, which is basically unused today.

**Comment by algernon** [comment]
While it is indeed easy to add type-hint support to patterndb, and not very hard to do the same for json-parser, the harder part is making sure that this information is available by the time the log message reaches the destination.

(at least, way back when, this was the problematic part of it)

**Comment by bazsi** [comment]
I dont see why it wouldn't reach it. It might not be returned by the
LogMessage accessors.

**Comment by faxm0dem** [comment]
Here's a little context on our use-case.

We're using Elasticsearch as backend for our logs, and use several syslog-ng parsers to extract nv-pairs.
Our patterndb is managed by puppet: users can publish new rules to production by pushing them to git. 

Unless a custom Elasticsearch mapping is also created syslog-ng will index all new nv-pairs as strings, and users won't be able to plot their extracted values as line graphs in kibana.

I think it would make sense that patterndb sets a default type-hint for each pattern parser, *e.g.* string for `@EMAIL@` or float for `@NUMBER`, but also allow for overriding that default, *e.g.* `@NUMBER:foo::type=string@`

**Comment by bazsi** [comment]
Hi,

Thanks for the explanation. I had exactly the same idea back in the day
I've implemented type support in db-parser(). There's one issue, that needs
some thinking: the type information is not always available at this point,
it is only available when we are using "indirect" references within our
LogMessage representation.

Indirects are used when db-parser() extracts a value from a longer string
(e.g. $MSG) and the extracted value is the same as the string in the
original value. e.g. if you parse something with @NUMBER@, that would still
be found in $MSG. In some cases the parsers can "convert" the value, where
indirect entries wouldn't be used.

This would still be an OKish constraint, however, if at a later point $MSG
gets changed, we would convert indirect entries to direct ones, losing the
type information. This constraint I consider a bit too limiting. I'll check
if there's an easy way out.
